teps/0137-cloudevents-controller.md

```diff
@@ -0,0 +1,527 @@
+used to collect the data required to sent the event. This approach allows for events to be
+sent at the start and completion of a Task only.
+
+An alternative implementation could be to provide Task authors with a file system based API,
+similar to what is available for results, accessible via variables, which might allow for
+events to be sent at any point during the execution of a Task. Such functionality would
+require continuous monitoring for the file-system pod to surface the data for the events
+controller to produce and send the final event.
+
+### Reusability
+
+The implementation of Tekton v1, v2 and CDEvents core events will be based on the existing
+code in `tektoncd/pipeline` and `tektoncd/experimental`.
+
+The support for additional CDEvents format may reuse the strategy in `tektoncd/experimental`
+or add new functionality depending on how the feature is designed.
+
+### Simplicity
+
+The proposed feature by default introduces minimal changes to the user experience and
+no changes for the Tekton operator.
+
+Operators will gain extra flexibility:
+- select the event format
+- tune RBAC permissions for the event producing component in Tekton
+
+### Flexibility
+
+This proposal introduces a dependency to the [CDEvents golang SDK][cdevents-go-sdk],
+which it turn depends on the `CloudEvents` SDK, which is already a dependency for Tekton,
+as well as on the [`package URL (pURL)` go SDK][purl-go-sdk].
+
+### Conformance
+
+There is no conformance requirement associated with events today, even though that's
+something we may consider in future. There adoption of CDEvents is an ideal choice
+for Tekton conformant system, as the ability to produce CDEvents brings interoperability
+value beyond the conformance to the Tekton API.
+
+### User Experience
+
+<!--
+(optional)
+
+Consideration about the user experience. Depending on the area of change,
+users may be Task and Pipeline editors, they may trigger TaskRuns and
+PipelineRuns or they may be responsible for monitoring the execution of runs,
+via CLI, dashboard or a monitoring system.
+
+Consider including folks that also work on CLI and dashboard.
+-->
+
+### Performance
+
+Moving CloudEvents to a separated controller reduces the functional
+responsibilities of the core controllers and it allows them and the
+event controllers to scale independently.
+
+The separated CloudEvents controller does not share the same context
+that the the main controllers have; because it must maintain a cache
+of the events that have been sent until now, to avoid event duplicates.
+The approach used today is to use a per-controller in-memory cache.
+The sharding of resources across controllers when scaling horizontally
+allows for the cache to be shared too and held in memory.
+
+A simple LRU cache is used in the current implementation. It shall be
+considered whether to switch to a [bette] performing cache library][go-cache-comparison].
```

Comment:
NIT: is there a typo? also this does not format well - https://github.com/tektoncd/community/blob/2a50cdbb4c4ea7f83a9e502d200aa84858b265f5/teps/0137-cloudevents-controller.md#performance

Reply:
Indeed, nice catch, thank you